docs/architecture/adr/0022-server-authorization.md

```diff
@@ -0,0 +1,244 @@
+  `CollectionUser`) should be considered their own resource and defined and checked separately
+
+To perform an authorization check:
+
+- call `AuthorizationService.AuthorizeAsync(user, resource, operations)`. It returns an
+  `AuthorizationResult` that indicates a success or failure
+- for a success result, at least 1 handler must have authorized the action, and no handler must have
+  expressly denied the action
+
+#### Advantages
+
+- included in ASP.NET - standard C# code, no additional dependencies
+- already partially in use for organization collections - we would be refining and then expanding
+  that use
+- handlers provide good encapsulation of authorization logic, separate to other concerns and the
+  `AuthorizationService` implementation itself
+- the fixed `AuthorizationService` interface (accepting only the user, the resource, and the
+  operations) enforces a consistent usage across our internal teams
+- teams can write and have code ownership over their own handlers
+- flexibly supports additional sources of authorization (e.g. scoped user API keys) by defining
+  additional handlers
+
+#### Disadvantages
+
+- performance:
+  - the `AuthorizationService` interface requires that the resources are fetched from the database
+    before an authorization decision can be made. This makes it impractical for large read
+    operations, such as syncing a user's vault, as we cannot realistically fetch _all_ ciphers from
+    the database and use `AuthorizationService` filter them. Therefore, some high frequency/high
+    cost operations would continue to use authorization logic embedded in database queries
+  - the authorization handlers may need to fetch additional data from the database to make a
+    decision. This may cause many database reads, particularly when iterating over multiple items,
+    unless an additional caching solution is implemented. This is probably manageable for now by
+    caching frequently needed data in `CurrentContext` or the handler itself, both of which are
+    scoped to the lifetime of the request
+- for the reasons above, it is not a complete solution: it would cover _most_ but not all of our use
+  cases
+- server-side solution only - client code (particularly in Admin Console) needs to continue to
+  maintain its own duplicate logic to determine what UI flows it should show/hide and enable/disable
+- a server endpoint needs to understand what resources are affected, which can be nuanced when
+  dealing with relational data (e.g. saving a vault item may require separate authorization checks
+  for the `Cipher` resource and the `CollectionCipher` relationship). However, this may be an issue
+  for other solutions as well, as our endpoints often update multiple resources at once
+
+### [OpenFGA](https://openfga.dev)
+
+#### Summary
+
+OpenFGA is an open-source implementation of Google's Zanzibar authorization system, which is used
```

Comment:
🎨 Link to https://research.google/pubs/zanzibar-googles-consistent-global-authorization-system/.